File: sample/cpp/hnsw_rabitq_indexing.cpp
#include <algorithm>
#include <cstdint>
#include <iostream>
#include <vector>

#include "rabitqlib/index/hnsw/hnsw.hpp"
#include "rabitqlib/utils/io.hpp"
#include "rabitqlib/utils/stopw.hpp"

using PID = rabitqlib::PID;
using index_type = rabitqlib::hnsw::HierarchicalNSW;
using data_type = rabitqlib::RowMajorArray<float>;
using gt_type = rabitqlib::RowMajorArray<uint32_t>;

int main(int argc, char* argv[]) {
    if (argc < 8) {
        std::cerr << "Usage: " << argv[0]
                  << " <arg1> <arg2> <arg3> <arg4> <arg5> <arg6> <arg7> <arg8>\n"
                  << "arg1: path for data file, format .fvecs\n"
                  << "arg2: path for centroids file, format .fvecs\n"
                  << "arg3: path for cluster_ids file, format .ivecs\n"
                  << "arg4: m (degree bound) for hnsw\n"
                  << "arg5: ef for indexing \n"
                  << "arg6: total number of bits for quantization\n"
                  << "arg7: path for saving index\n"
                  << "arg8: metric type (\"l2\" or \"ip\"), l2 by default\n"
                  << "arg9: if use faster quantization (\"true\" or \"false\"), false by "
                     "default\n";
        exit(1);
    }

    char* data_file = argv[1];
    char* centroid_file = argv[2];
    char* cid_file = argv[3];
    size_t m = atoi(argv[4]);
    size_t ef = atoi(argv[5]);
    size_t total_bits = atoi(argv[6]);
    char* index_file = argv[7];

    rabitqlib::MetricType metric_type = rabitqlib::METRIC_L2;
    if (argc > 8) {
        std::string metric_str(argv[8]);
        if (metric_str == "ip" || metric_str == "IP") {
            metric_type = rabitqlib::METRIC_IP;
        }
    }
    if (metric_type == rabitqlib::METRIC_IP) {
        std::cout << "Metric Type: IP\n";
    } else if (metric_type == rabitqlib::METRIC_L2) {
        std::cout << "Metric Type: L2\n";
    }

    bool faster_quant = false;
    if (argc > 9) {
        std::string faster_str(argv[9]);
        if (faster_str == "true") {
            faster_quant = true;
            std::cout << "Using faster quantize for indexing...\n";
        }
    }

    data_type data;
    data_type centroids;
    gt_type cluster_id;

    rabitqlib::load_vecs<float, data_type>(data_file, data);
    rabitqlib::load_vecs<float, data_type>(centroid_file, centroids);
    rabitqlib::load_vecs<uint32_t, gt_type>(cid_file, cluster_id);

    size_t num_points = data.rows();
    size_t dim = data.cols();

    size_t random_seed = 100;  // by default 100
    auto* hnsw = new rabitqlib::hnsw::HierarchicalNSW(
        num_points, dim, total_bits, m, ef, random_seed, metric_type
    );

    // Build the index with the first 80% of the data, then stream the
    // remaining 20% in via add() in 10%-of-dataset batches to exercise the
    // dynamic-add path.
    size_t build_count = (num_points * 80) / 100;
    size_t batch_size = num_points / 10;  // 10% of the dataset per batch
    if (batch_size == 0) {
        batch_size = 1;
    }

    rabitqlib::StopW stopw;
    stopw.reset();

    hnsw->construct(
        centroids.rows(),
        centroids.data(),
        build_count,
        data.data(),
        cluster_id.data(),
        0,
        faster_quant
    );

    float build_time = stopw.get_elapsed_micro() / 1e6;
    std::cout << "built index with " << build_count << " / " << num_points
              << " points in " << build_time << "s" << '\n';

    // Labels mirror the global row index so they stay consistent with the
    // built portion (construct uses the row index as the label).
    std::vector<PID> labels(num_points);
    for (size_t i = 0; i < num_points; i++) {
        labels[i] = static_cast<PID>(i);
    }

    stopw.reset();
    for (size_t start = build_count; start < num_points; start += batch_size) {
        size_t end = std::min(start + batch_size, num_points);
        // num_threads = 0 -> use all available cores for the batch insert.
        hnsw->add(data.data() + (start * dim), labels.data() + start, end - start, 0);
        std::cout << "added points [" << start << ", " << end << ")\n";
    }
    float add_time = stopw.get_elapsed_micro() / 1e6;

    std::cout << "added " << (num_points - build_count) << " points in " << add_time
              << "s" << '\n';
    std::cout << "total indexing time = " << (build_time + add_time) << "s" << '\n';
    hnsw->save(index_file);

    std::cout << "index saved..." << '\n';

    return 0;
}

File: sample/python/hnsw_rabitq_indexing.py
import argparse
from time import time

from rabitqlib import HnswIndex

from utils import cluster_data, read_fvecs

# ──────────────────────────────────────────────
# Default configuration
# ──────────────────────────────────────────────
NUM_CLUSTERS = 256  # number of clusters
M = 16  # degree bound for HNSW
EF_CONSTRUCTION = 200  # ef for indexing
TOTAL_BITS = 8  # total number of bits for quantization
METRIC = "l2"  # "l2" or "ip"
FASTER_QUANT = False  # use faster quantization
NUM_THREADS = 16  # number of threads for build
# ──────────────────────────────────────────────


def main(args=None) -> None:
    # 1. Load data
    data = read_fvecs(args.data_file)
    print(f"Data shape: {data.shape}")

    # 2. Cluster with FAISS
    centroids, cluster_ids = cluster_data(
        data, args.num_clusters, args.metric, args.num_threads
    )
    print(f"Centroids: {centroids.shape}, cluster_ids: {cluster_ids.shape}")

    # 3. Build HNSW index
    n, dim = data.shape
    if args.base_bits is None:
        print(f"\nBuilding HNSW index: n={n}, dim={dim}, M={args.degree}, "
              f"ef={args.ef_construction}, bits={args.total_bits}, metric={args.metric}")
    else:
        print(f"\nBuilding HNSW index: n={n}, dim={dim}, M={args.degree}, "
              f"ef={args.ef_construction}, base_bits={args.base_bits}, "
              f"extra_bits={args.extra_bits or 0}, metric={args.metric}")

    idx = HnswIndex(
        dim=dim,
        max_elements=n,
        M=args.degree,
        ef_construction=args.ef_construction,
        nbits=args.total_bits,
        metric=args.metric,
        base_bits=args.base_bits,
        extra_bits=args.extra_bits,
    )

    t0 = time()
    idx.build(
        data,
        centroids,
        cluster_ids,
        num_threads=args.num_threads,
        fast_quantization=args.faster_quant,
    )
    print(f"Indexing time: {time() - t0:.2f}s")

    idx.save(args.index_file)
    print(f"Index saved → {args.index_file}")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="RaBitQ HNSW Manager")

    parser.add_argument("data_file", type=str, help="Path to the data file")
    parser.add_argument("index_file", type=str, help="Path to save the index")
    parser.add_argument(
        "--num-clusters",
        dest="num_clusters",
        type=int,
        metavar="INT",
        default=256,
        help="Number of clusters for quantization",
    )
    parser.add_argument(
        "--degree",
        dest="degree",
        type=int,
        metavar="INT",
        default=M,
        help="Degree bound for HNSW",
    )
    parser.add_argument(
        "--ef-construction",
        dest="ef_construction",
        type=int,
        metavar="INT",
        default=EF_CONSTRUCTION,
        help="EF parameter for index construction",
    )
    parser.add_argument(
        "--total-bits",
        dest="total_bits",
        type=int,
        metavar="INT",
        default=TOTAL_BITS,
        help="Total number of bits for quantization",
    )
    parser.add_argument("--base-bits", dest="base_bits", type=int, metavar="INT", default=None, help="Base bits for the new x+y quantization mode (1-8). Leave unset to use --total-bits with the classic 1-bit-base scheme")
    parser.add_argument("--extra-bits", dest="extra_bits", type=int, metavar="INT", default=None, help="Extra/refinement bits for x+y quantization (0-8, default 0). Only used when --base-bits is set")
    parser.add_argument(
        "--metric",
        dest="metric",
        type=str,
        default="l2",
        choices=["l2", "ip"],
        help="Distance metric (l2 or ip)",
    )
    parser.add_argument(
        "--faster-quant",
        dest="faster_quant",
        action="store_true",
        help="Use faster quantization method",
    )
    parser.add_argument(
        "--num-threads",
        dest="num_threads",
        type=int,
        metavar="INT",
        default=NUM_THREADS,
        help="Number of threads for building the index",
    )

    args = parser.parse_args()

    main(args)
